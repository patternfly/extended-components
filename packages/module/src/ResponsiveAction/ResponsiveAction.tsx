import React from 'react';
import { ButtonProps } from '@patternfly/react-core';

export interface ResponsiveActionProps extends ButtonProps {
  /** Determines whether the action should be displayed next to dropdown if possible */
  isPinned?: boolean;
  /** Determines whether the action should always be displayed as pinned */
  isPersistent?: boolean;
  /** Key for the action */
  key?: string;
  /** Action label */
  children: React.ReactNode;
};

// This component is only used declaratively - rendering ishandled by ResponsiveActions
export const ResponsiveAction: React.FunctionComponent<ResponsiveActionProps> = (_props: ResponsiveActionProps) => null;

export default ResponsiveAction;