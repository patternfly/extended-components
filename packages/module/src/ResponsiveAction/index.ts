export { default } from './ResponsiveAction';
export * from './ResponsiveAction';
