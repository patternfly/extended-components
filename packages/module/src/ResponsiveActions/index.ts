export { default } from './ResponsiveActions';
export * from './ResponsiveActions';
