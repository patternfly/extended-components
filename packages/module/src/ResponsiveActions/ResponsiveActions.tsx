import React, { useState } from 'react';
import { Button, Dropdown, DropdownList, MenuToggle, OverflowMenu, OverflowMenuContent, OverflowMenuControl, OverflowMenuDropdownItem, OverflowMenuGroup, OverflowMenuItem, OverflowMenuProps } from '@patternfly/react-core';
import { EllipsisVIcon } from '@patternfly/react-icons';
import { ResponsiveActionProps } from '../ResponsiveAction';

export interface ResponsiveActionsProps extends Omit<OverflowMenuProps, 'ref' | 'breakpoint'> {
  /** Indicates breakpoint at which to switch between horizontal menu and vertical dropdown */
  breakpoint?: OverflowMenuProps['breakpoint'];
  /** Custom OUIA ID */
  ouiaId?: string;
  /** Child actions to be displayed */
  children: React.ReactNode;
}

export const ResponsiveActions: React.FunctionComponent<ResponsiveActionsProps> = ({ ouiaId = 'ResponsiveActions', breakpoint = 'lg', children, ...props }: ResponsiveActionsProps) => {
  const [ isOpen, setIsOpen ] = useState(false);

  // separate persistent, pinned and collapsed actions
  const persistentActions: React.ReactNode[] = [];
  const pinnedActions: React.ReactNode[] = [];
  const dropdownItems: React.ReactNode[] = [];

  React.Children.forEach(children, (child, index) => {
    if (React.isValidElement<ResponsiveActionProps>(child)) {
      const { isPersistent, isPinned, key = index, children, onClick, ...actionProps } = child.props;

      if (isPersistent || isPinned) {
        (isPersistent ? persistentActions : pinnedActions).push(
          <OverflowMenuItem key={key} isPersistent={isPersistent}>
            <Button onClick={onClick} ouiaId={`${ouiaId}-action-${key}`} {...actionProps}>
              {children}
            </Button>
          </OverflowMenuItem>
        );
      }
      if (!isPersistent) {
        dropdownItems.push(
          <OverflowMenuDropdownItem key={key} onClick={onClick} isShared={isPinned} ouiaId={`${ouiaId}-action-${key}`}>
            {children}
          </OverflowMenuDropdownItem>
        );
      }
    }
  });

  return (
    <OverflowMenu breakpoint={breakpoint} data-ouia-component-id={`${ouiaId}-menu`} {...props}>
      {persistentActions.length > 0 ? (
        <OverflowMenuContent isPersistent data-ouia-component-id={`${ouiaId}-menu-persistent-content`}>
          <OverflowMenuGroup groupType="button" data-ouia-component-id={`${ouiaId}-menu-persistent-group`} isPersistent>
            {persistentActions}
          </OverflowMenuGroup>
        </OverflowMenuContent>
      ) : null}
      {pinnedActions.length > 0 ? (
        <OverflowMenuContent data-ouia-component-id={`${ouiaId}-menu-pinned-content`}>
          <OverflowMenuGroup groupType="button" data-ouia-component-id={`${ouiaId}-menu-pinned-group`}>
            {pinnedActions}
          </OverflowMenuGroup>
        </OverflowMenuContent>
      ) : null}
      {dropdownItems.length > 0 && (
        <OverflowMenuControl hasAdditionalOptions data-ouia-component-id={`${ouiaId}-menu-control`}>
          <Dropdown
            ouiaId={`${ouiaId}-menu-dropdown`}
            onSelect={() => setIsOpen(false)}
            toggle={(toggleRef) => (
              <MenuToggle
                ouiaId={`${ouiaId}-menu-dropdown-toggle`}
                ref={toggleRef}
                aria-label="Actions overflow menu"
                variant="plain"
                onClick={() => setIsOpen(!isOpen)}
                isExpanded={isOpen}
              >
                <EllipsisVIcon />
              </MenuToggle>
            )}
            isOpen={isOpen}
            onOpenChange={setIsOpen}
          >
            <DropdownList data-ouia-component-id={`${ouiaId}-menu-dropdown-list`}>
              {dropdownItems}
            </DropdownList>
          </Dropdown>
        </OverflowMenuControl>
      )}
    </OverflowMenu>
  );
};

export default ResponsiveActions;