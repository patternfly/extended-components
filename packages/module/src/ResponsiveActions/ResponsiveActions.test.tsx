import React from 'react';
import { render } from '@testing-library/react';
import ResponsiveActions from './ResponsiveActions';
import ResponsiveAction from '../ResponsiveAction';

describe('ResponsiveActions component', () => {
  describe('should render correctly', () => {

    test('ResponsiveActions', () => {
      const { container } = render(
        <ResponsiveActions breakpoint="lg">
          <ResponsiveAction isPersistent>Persistent action</ResponsiveAction>
          <ResponsiveAction isPinned variant='secondary'>Pinned action</ResponsiveAction>
          <ResponsiveAction>Overflow action</ResponsiveAction>
        </ResponsiveActions>);
      expect(container).toMatchSnapshot();
    });
  });
});