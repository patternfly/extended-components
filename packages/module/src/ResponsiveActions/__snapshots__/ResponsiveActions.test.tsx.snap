// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponsiveActions component should render correctly ResponsiveActions 1`] = `
<div>
  <div
    class="pf-v5-c-overflow-menu"
    data-ouia-component-id="ResponsiveActions-menu"
  >
    <div
      class="pf-v5-c-overflow-menu__content"
    >
      <div
        class="pf-v5-c-overflow-menu__group pf-m-button-group"
        data-ouia-component-id="ResponsiveActions-menu-persistent-group"
      >
        <div
          class="pf-v5-c-overflow-menu__item"
        >
           
          <button
            aria-disabled="false"
            class="pf-v5-c-button pf-m-primary"
            data-ouia-component-id="ResponsiveActions-action-0"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            type="button"
          >
            Persistent action
          </button>
           
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-overflow-menu__content"
    >
      <div
        class="pf-v5-c-overflow-menu__group pf-m-button-group"
        data-ouia-component-id="ResponsiveActions-menu-pinned-group"
      >
        <div
          class="pf-v5-c-overflow-menu__item"
        >
           
          <button
            aria-disabled="false"
            class="pf-v5-c-button pf-m-secondary"
            data-ouia-component-id="ResponsiveActions-action-1"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            type="button"
          >
            Pinned action
          </button>
           
        </div>
      </div>
    </div>
    <div
      class="pf-v5-c-overflow-menu__control"
      data-ouia-component-id="ResponsiveActions-menu-control"
    >
       
      <button
        aria-expanded="false"
        aria-label="Actions overflow menu"
        class="pf-v5-c-menu-toggle pf-m-plain"
        data-ouia-component-id="ResponsiveActions-menu-dropdown-toggle"
        data-ouia-component-type="PF5/MenuToggle"
        data-ouia-safe="true"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="pf-v5-svg"
          fill="currentColor"
          height="1em"
          role="img"
          viewBox="0 0 192 512"
          width="1em"
        >
          <path
            d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
          />
        </svg>
      </button>
       
    </div>
  </div>
</div>
`;
